plugins/scale/wayfire/plugins/scale-transform.hpp

```diff
@@ -0,0 +1,282 @@
+#pragma once
+
+#include <wayfire/view-transform.hpp>
+#include <wayfire/nonstd/observer_ptr.h>
+#include <wayfire/render-manager.hpp>
+#include <string>
+#include <list>
+#include <algorithm>
+
+namespace wf
+{
+/**
+ * transformer used by scale -- it is an extension of the 2D transformer
+ * with the ability to add overlays
+ */
+class scale_transformer_t : public wf::view_2D
+{
+  public:
+    scale_transformer_t(wayfire_view view) : wf::view_2D(view)
+    {}
+    ~scale_transformer_t()
+    {}
+
+    struct padding_t
+    {
+        unsigned int top    = 0;
+        unsigned int left   = 0;
+        unsigned int bottom = 0;
+        unsigned int right  = 0;
+
+        /**
+         * Compare paddings. Returns true if ANY of the dimensions in this
+         * is smaller than other. Note that this does not define an ordering
+         * on paddings.
+         */
+        bool any_smaller_than(const padding_t& other) const
+        {
+            return (top < other.top) || (left < other.left) ||
+                   (bottom < other.bottom) || (right < other.right);
+        }
+
+        /**
+         * Helper function to extend padding to ensure it is at least as
+         * large as other.
+         */
+        void extend(const padding_t& other)
+        {
+            top    = std::max(top, other.top);
+            left   = std::max(left, other.left);
+            bottom = std::max(bottom, other.bottom);
+            right  = std::max(right, other.right);
+        }
+    };
+
+    uint32_t get_z_order() override
+    {
+        /* ensure that scale comes after fire */
+        return wf::TRANSFORMER_HIGHLEVEL + 2;
+    }
+
+    /**
+     * Effect hook called in the following circumstances:
+     *  (1) scale's animation is running
+     *  (2) another transformer changed the view's size, so the overlay and
+     * padding might need to be updated
+     *
+     * The second case can happen just during rendering the view (from
+     * view::render_transformed()), so it should not call damage.
+     *
+     * @return Whether the overlay changed. In this case damage will be
+     * scheduled for the view.
+     *
+     * Damage will also be scheduled if the padding's size changed.
+     */
+    using pre_hook_t = std::function<bool (void)>;
+
+    /**
+     * Render hook for drawing on top of the surface after the transform, during
+     * rendering the view.
+     *
+     * @param fb     The framebuffer to render to.
+     * @param damage Damaged region to render.
+     */
+    using render_hook_t = std::function<void (const wf::framebuffer_t& fb,
+        const wf::region_t& damage)>;
+
+    /**
+     * Overlays that can be added to this transformer. Hooks are called
+     * similarly to render-manager.
+     */
+    struct overlay_t : public wf::custom_data_t
```

Comment:
why does this need to derive from custom_data_t ?

Reply:
You're right, it can have its own virtual destructor; I'll change it.